### DISCUSSION TITLE
Multiple squeezelite instances, bluetooth speakers

### DISCUSSION THREAD

**Discussion by barrelltitor** [GioF71/squeezelite-docker#242] [General]
Hey, I'm trying to run two squeezelite instances on the same machine that has two connected Bluetooth speakers.

I tried setting the audio device as such - SQUEEZELITE_AUDIO_DEVICE=bluez_sink.44_07_0B_44_CB_15.a2dp_sink for one of them, and with the other MAC for the other squeezelite container

Both of them are playing through the exact same sink, and the logs show that -o wasn't added:

[/app/bin/squeezelite-pulseaudio -s 192.168.88.196 -n "Kitchen Speaker" -C 2 -r 44100 -e dsd]

Any idea what I'm doing wrong? Here's my full compose:

`version: "3"

services:
  squeezelite-my-usb-dac:
    image: giof71/squeezelite:stable #tried latest as well
    container_name: kitchen-speaker
    privileged: true
    volumes:
      - /run/user/1000/pulse:/run/user/1000/pulse
    cap_add:
      - NET_RAW
      - NET_ADMIN      
    environment:
      - SQUEEZELITE_MODE=PULSE
      - SQUEEZELITE_NAME=Kitchen Speaker
      - SQUEEZELITE_SERVER_PORT=192.168.88.196
      - SQUEEZELITE_AUDIO_DEVICE=bluez_sink.44_07_0B_44_CB_15.a2dp_sink
    restart: unless-stopped
    `

**Comment by GioF71** [comment]
Hello, you are using SQUEEZELITE_MODE=PULSE so the variable SQUEEZELITE_AUDIO_DEVICE has no effect here.
You can use a tool like PulseAudio Volume Control to associate the application to the preferred output device.
I don't know if the pulse binary allows to select a preferred device, I need to investigate.

As an alternative, you can try using alsa mode, but I don't have any experience about this with bluetooth devices, and if you are running a desktop the benefit would be limited. In any case, can you post the result of 

`cat /proc/asound/cards`
and

`aplay -l`

when the bt speakers are connected?

**Comment by GioF71** [comment]
Also maybe try to run:

`docker exec -it kitchen-speaker bash`

then try to run:

`/app/bin/squeezelite-pulseaudio -l`

so we can see if some devices are listed.

**Comment by barrelltitor** [comment]
I'm on a Raspberry Pi 3B+ without a desktop environment
I have tried using alsa with bluezalsa. I compiled bluealsa from source as it was not available for the PI3B+, I installed squeezelite as well and running it on the host OS worked like a charm with -o "bluealsa:DEV=F0:EF:86:C0:12:5C,PROFILE=a2dp,SRV=org.bluealsa" 

Then I tried 
- SQUEEZELITE_AUDIO_DEVICE=bluealsa:DEV=F0:EF:86:C0:12:5C,PROFILE=a2dp,SRV=org.bluealsa

Which did not work. I found in random github issues that bluealsa might need installed in the container so its alsa libraries are available to be used.

However, the error I got was 
```playback open error: No such file or directory```
Maybe I missed mounting something? 

The container is privileged, and I also tried mounting the following:

  volumes:
      - /etc/asound.conf:/etc/asound.conf:ro
      - /usr/lib/arm-linux-gnueabihf/alsa-lib:/usr/lib/arm-linux-gnueabihf/alsa-lib:ro
      - /run/dbus/system_bus_socket:/run/dbus/system_bus_socket
 Also tried mounting /var/run/dbus
 
 In all these cases the same playback error appeared when trying to start up the container, and it kept rebooting every time this error appeared
 
 I'm going to get another SD card to setup pulseaudio on it so I can more easily test without having to install/reinstall things
 
 aplay -l will only show the hardware connected devices, it won't show anything that's in bluetooth. To list bluetooth devices you usually need to use bluealsa or pulseaudio's pactl
 
 I tried setting the sink with pactl but squeezelite did not appear in there, maybe I'll need to actually try playing something to properly see it

**Comment by GioF71** [comment]
Ok understood, I assumed it was a desktop because I saw that SQUEEZELITE_MODE=PULSE.
What you want to do is quite complicated and mostly unknown territory for me.
I stumbled upon a stack overflow: https://raspberrypi.stackexchange.com/questions/90267/how-to-stream-sound-to-a-bluetooth-device-from-a-raspberry-pi-zero
If you are able to instruct aplay to play to the named device "mid" like in the example using /etc/asound.conf, you will _probably_ be able to use that alsa device directly (without all the bluetooth complications), of course mounting the asound.conf file to the container. We would be leaving the complex configurations on the host device.

Another repository I found is [this one](https://github.com/oweitman/squeezelite-bluetooth) but it looks a bit outdated, last commit is 3 years ago. Also it would look very complicated to bring all of this in the container image.

About your test, when you used 
`SQUEEZELITE_AUDIO_DEVICE=bluealsa:DEV=F0:EF:86:C0:12:5C,PROFILE=a2dp,SRV=org.bluealsa`
what was the resulting command line?

**Comment by barrelltitor** [comment]
I got aplay playing by default to the bluetooth speaker when just doing `aplay a.wav`

The container unfortunately crashes after checking if the variables are specified:


```
Variable SQUEEZELITE_VISUALIZER not specified

User mode disabled

Command Line: [/app/bin/squeezelite -s 192.168.88.196 -n "Living Room Speaker" -C 2 -D 500]

[15:09:50.992862] test_open:281 playback open error: No such device or address

[15:09:50.996502] output_init_common:401 unable to open output device: default
```

This is my compose:

```version: "3"

services:
  squeezelite-my-usb-dac:
    image: giof71/squeezelite:latest
    container_name: squeezelite
    privileged: true
    devices: 
      - /dev/snd:/dev/snd
    volumes:
      - /home/pi/.asoundrc:/etc/asound.conf:ro
      #- /usr/lib/arm-linux-gnueabihf/alsa-lib:/usr/lib/arm-linux-gnueabihf/alsa-lib:ro  # Tried this too with no luck
    cap_add:
      - NET_RAW
      - NET_ADMIN      
    environment:
      - SQUEEZELITE_NAME=Living Room Speaker
      - AUDIO_GID=29 # tried with this on/off as well as with the docker user: 1000:29 option
      - SQUEEZELITE_SERVER_PORT=192.168.88.196
    restart: unless-stopped
    ```
    
    Any ideas?

**Reply by GioF71** [reply]
uhm, what is the content of:
`/home/pi/.asoundrc`
?
Maybe you can set the output (SQUEEZELITE_AUDIO_DEVICE) to the named device defined there...

**Reply by GioF71** [reply]
If setting SQUEEZELITE_AUDIO_DEVICE does not work, maybe some libraries must be installed on the container.
Do you have the list of what you installed on the host, and know what is needed to play to that device? If so, you can try:

docker exec -it kitchen-speaker bash

then

```
apt-get update
apt-get install <those packages>

```
exit the shell, restart (without recreating) the container

**Comment by GioF71** [comment]
Hello, did this work?
I read your posts with more attention, I overlooked that you had to compile bluealsa.
Is there a guide or some online resource to describe this process? Or, does any distro include compiled versions of bluealsa?

Thank you

**Reply by barrelltitor** [reply]
I didn't have time to try again unfortunately, but I am very interested in this. I could successfully connect all my bluetooth speakers to a single pi and manually play to all of them, so being able to have a dedicated squeezelite pi would be great! 

I also managed to get multiple squeezelite instances on the same pi with no issues, all that's left is to try to build bluealsa

The problem is that if I use bluealsa instead of pulse, I cannot get the container to actually start. I'll try to install both at once or plug in a wired speaker to test and will report back in some days

**Comment by barrelltitor** [comment]
Got some more time to try this out and I have some great news, got it working without needing to custom compile anything! 

After running ```apt -y update && apt -y install bluetooth bluez-alsa-utils alsa-utils``` inside the container, I was able to aplay(reason for installing alsa-utils) a wav file. Then I ran /app/bin/squeezelite -s 192.168.88.196 -n "Living Room Speaker" -C 2 -D 500 and it worked perfectly fine!

Changing my /etc/asound.conf inside the container used a different bluetooth speaker!

I had a bunch of stuff mounted, but after some more testing this works fine:

 ```
sudo docker run -v ./lrs:/etc/asound.conf -v /run/dbus/system_bus_socket:/run/dbus/system_bus_socket  -it --entrypoint /bin/bash giof71/squeezelite:latest

 apt -y update && apt -y install bluetooth bluez-alsa-utils && /app/bin/squeezelite -s 192.168.88.196 -n "Living Room Speaker" -C 2 -D 500 
```

I did not manage to test multiple instances at once with different speakers, that's next, but I don't have time anymore today, maybe someone else can try it out

I am currently running this with a custom image and it's working flawlessly 
```
FROM giof71/squeezelite:latest
RUN apt -y update && apt -y install bluetooth bluez-alsa-utils
```


EDIT:

Here's the compose too

```
version: "3"

services:
  squeezelite:
    image: squeezeblue # custom built from above
    container_name: squeezelite
    volumes:
      - /home/pi/audio/bathroomspk:/etc/asound.conf:ro
      - /run/dbus/system_bus_socket:/run/dbus/system_bus_socket
    cap_add:
      - NET_RAW # not sure if these two are needed
      - NET_ADMIN # not sure if these two are needed      
    environment:
      - SQUEEZELITE_NAME=Bathroom Speaker
      - SQUEEZELITE_SERVER_PORT=192.168.88.196
    restart: unless-stopped
    
```

**Comment by GioF71** [comment]
Hello, nice to hear that. What would be the content of that asound.conf file?

Also, how can we support this configuration? The only thing I would not like to do is include these packages in the main build. But I can maybe offer the possibility to build a custom image using build.sh which will include the packages.
Also a documentation page (maybe in a discussion) might be nice.

What do you think?

**Comment by barrelltitor** [comment]
I don't mind now as I already make a bunch of custom images, but for users not very experienced with Docker it might be better to have a bluetooth tag, e.g giof71/squeezelite:latest-bluetooth
I knew I forgot to include something! 

The asound.conf file 
```
 pcm.!default {
        type plug
        slave {
            pcm {
                type bluealsa
                device 08:EB:ED:64:BC:C4
                profile "a2dp"
            }
        }
        hint {
            show on
            description "BathroomSpeaker"
        }
     }

```

You just need to replace the MAC with the one of your specific Speaker. I don't think the hint is actually necessary, that could be removed

**Comment by GioF71** [comment]
Ok I will try it myself then write something, maybe I can also add the image to the builds. We'll see
Thank you!

**Comment by barrelltitor** [comment]
Just tested with 2 speakers on a single pi3B+ on wifi using its built in bluetooth and it works flawlessly. Let me know if I can help with testing, documentation or anything else!

Here's an image with the above changes if anyone wants to try it out ghcr.io/barrelltitor/squeezeblue though I'd recommend trying to build it yourself to learn, it's really easy!

**Comment by GioF71** [comment]
Hello, thank you. I will try asap

**Comment by GioF71** [comment]
Hello, sorry for the long delay. The main problem is that I don't have a lot of bluetooth devices... just one [crappy usb key](https://www.amazon.it/gp/product/B0C1VLZ1K9/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) and a Pi3 with moode audio installed working also as a bluetooth renderer. But that is in the living room. So it's quite difficult to make progress for me.
However, can you have a look at [there](https://hub.docker.com/r/giof71/squeezelite/tags?page=1&name=f277b1) images?
If you set the new variable INSTALL_BLUETOOTH_LIBRARIES to yes, and avoid to run as a user with eg "--user 1000:1000" on docker run or "user: 1000:1000" with docker-compose, the bluetooth libraries should be installed.
Let me know if this works, thank you

**Comment by barrelltitor** [comment]
Works, though having to install the libraries on every container (re)creation isn't really ideal. I have 2 of these speakers on the same device and will try getting a third one, so an update will result in 3 installs for the same files on a single machine. 

Would prefer to avoid unnecessary operations on the microsd and the ability to recreate the container without internet access

**Comment by barrelltitor** [comment]
I'm also getting quite a lot of unhandled grfe in the logs, not sure what that's from though, I think I've seen that error before switching to bluetooth as well

**Comment by GioF71** [comment]
Ok... let me think about a solution. I currently have "vanilla" builds, which install squeezelite from the repos, and another set of builds which use sourceforge binaries. I added lately the r2 variant, mostly useless as that functionality is included with the sourceforge versions.
Which base version are you using?
As an alternative, would you be fine building your libraries, and maybe pushing them to your own privare docker registry, so they are ready for your devices?

About you latest post, what do you mean by "unhandled grfe" ? Not sure :-)

**Comment by GioF71** [comment]
Hello, I have created a build for bluetooth, see [here](https://hub.docker.com/repository/docker/giof71/squeezelite/tags?page=1&ordering=last_updated&name=alsa-bt)
Can you test it? If successful I will include the build as one of the regulars.
Thank you

**Comment by barrelltitor** [comment]
Really sorry for the delayed reply! I'm happy to build my own docker images, yeah!

The one I tested before was giof71/squeezelite:feature-f277b1-debian-alsa-squeezelite-1.9.9

The new one works too, however both of them seem to give me (over time) tens of thousands of these errors 
![firefox_G2lH27Bv1D](https://github.com/GioF71/squeezelite-docker/assets/108460132/c03d612d-6049-4b8d-8d9d-7ee09987467c)
here's from a restart just now
![firefox_F69I47Nkuf](https://github.com/GioF71/squeezelite-docker/assets/108460132/97489472-2df5-4f44-b032-768efe4b7a1c)

They don't seem to be causing any issues, and I'm not sure if this is a squeezelite issue rather than with the container

**Comment by GioF71** [comment]
The image with bluetooth libraries is this one: giof71/squeezelite:feature-f277b2-debian-alsa-bt-squeezelite-1.9.9
But if the alsa version you used works, maybe those libraries aren't actually needed in the container...
I don't know about that log, we should have a look a the code at least to understand what triggers it

**Reply by barrelltitor** [reply]
You are misunderstanding. The libraries are indeed required. 

I was simply explaining what version I switched **from**, which is the one you told me a bit above  https://github.com/GioF71/squeezelite-docker/discussions/242#discussioncomment-8959079  https://hub.docker.com/r/giof71/squeezelite/tags?page=1&name=f277b1 using the environment variable INSTALL_BLUETOOTH_LIBRARIES
 
**At the moment** I am using the new one, giof71/squeezelite:feature-f277b2-debian-alsa-bt-squeezelite-1.9.9 and it does work, the only issue being that it spams the warnings

**Reply by GioF71** [reply]
Ok cool, thank you for explaining! So the builds are ok, I will promote to the main branch later today. Let's hope the proto issue can be solved somehow. Thank you!

**Comment by GioF71** [comment]
About the warning... here's where it's created: 
https://github.com/ralph-irving/squeezelite/blob/3cbbc535f3431880b54c78f5f9489484e4187cf0/slimproto.c#L537

I am not sure about the version.
Can you post your compose file (or docker run command)?

**Reply by barrelltitor** [reply]
Here's my current compose

```
version: "3"

services:
  squeezelite:
    image: giof71/squeezelite:feature-f277b2-debian-alsa-bt-squeezelite-1.9.9
    container_name: squeezelite
    volumes:
      - /home/pi/audio/lrs:/etc/asound.conf:ro
      - /run/dbus/system_bus_socket:/run/dbus/system_bus_socket
    cap_add:
      - NET_RAW # not sure if these two are needed
      - NET_ADMIN # not sure if these two are needed      
    environment:
      - SQUEEZELITE_NAME=Living Room Speaker
      - SQUEEZELITE_SERVER_PORT=192.168.88.196
      - SQUEEZELITE_MAC_ADDRESS=02:42:ac:12:00:02
    restart: unless-stopped
```

**Reply by barrelltitor** [reply]
I'll open an issue on the slimproto repo though since this is from their side not yours, thanks for looking into it <3

**Reply by GioF71** [reply]
Can you please add a reference to the ticket here, when you create it?
Thank you

**Reply by barrelltitor** [reply]
Will test it out tomorrow both in docker and on the pi itself, I'm pretty sure I got this when running squeezelite on the host itself as well when I tried yesterday but I didn't save the logs

**Reply by GioF71** [reply]
Ok, will be waiting.
In the meantime the change has been merged into main. You can then use this tag: debian-alsa-bt-squeezelite-current
I will probably get rid of that postfix "-squeezelite-current" in the tag, it's probably quite meaningless.

**Reply by barrelltitor** [reply]
Seems to be a harmless warning according to the music assistant developer, though I made the issue as I'm curious what is going on.  https://github.com/ralph-irving/squeezelite/issues/219

For now I just did
```
logging:
      driver: none
```
to avoid the spam

**Reply by GioF71** [reply]
Ok good to hear @barrelltitor, would you like to send in a pull request with a small page in markdown explaining how to configure bluetooth?

**Reply by barrelltitor** [reply]
Sure! Should I add it in the examples section in the README?

**Reply by GioF71** [reply]
I would ask you to write a separata Page and  add a link in the readme. This is because the readme is very close to the 25k limit for docker hub. The text would be truncated.
Actually I need to write a small readme for docker hub with the link to the GitHub repo.
But I need some time for that.
What this is done we will be ok to integrate your description in the main readme.

**Reply by GioF71** [reply]
Examples are welcome of course!

**Reply by barrelltitor** [reply]
Will do it tomorrow, don't have the time today unfortunately.  Will also add the instructions for getting everything running on the host as well as examples on how to auto connect the speaker on startup.

I really appreciate all the help, looks like everything is now sorted out!

**Reply by GioF71** [reply]
Of course. There's no rush!
